### DISCUSSION TITLE
UTF-8 issue with LiquidSoap

### DISCUSSION THREAD

**Discussion by SpeedyCharly** [savonet/liquidsoap#2155] [Q&A]
From; https://github.com/AzuraCast/AzuraCast/discussions/4925

A while back I asked about why UTF-8 doesn't recognize standard French characters in the metadata as seen here;

------------------------------
2021/12/25 09:24:18 [camomile:3] Failed to convert "Fetes - Yvan- V\233ronique en ondes" from UTF-8 to UTF-8 (CamomileLibrary__CharEncoding.Configure(Config).Malformed_code)!
2021/12/25 09:24:18 [input_streamer:3] New metadata chunk ? -- Fetes - Yvan- V?ronique en ondes.
-------------------------------

At the same time we poll the metadata from our Shoutcast stream [/7.html] to feed our chatroom bot which shows up properly:
![Chat](https://user-images.githubusercontent.com/66107794/149273831-1c393ab3-9c62-4e92-a1df-539a5a252a48.jpg)

Didn't really get a response the first time as it was embedded into another topic...

So, the question is; what does Shoutcast do that LS doesn't?

Could the problem reside with the Unicode library used?
[Camomile is a Unicode library for objective caml (a. k. a. OCaml or O'Caml) language. Camomile provides Unicode character type, UTF-8, UTF-16, UTF-32 strings, conversion to/from about 200 encodings, collation and locale-sensitive case mappings, and more. The library is currently designed for Unicode Standard 3.2.]

-----------------------------------

From; https://github.com/AzuraCast/AzuraCast/issues/4937

Reported before on difficulties with UTF-8 and LS supporting French custom characters. At the time, I also noted that the very same metadata was transcribed correctly in ShoutCast. Yes, we're on Shout, not Ice; this to accommodate our djs.

Thus, I thought that switching autoDJ from UTF-8 to ISO-8859-1 (going full Legacy) would solve the problem, as AzuraCast documentation seems to indicate.
![AZURA Module](https://user-images.githubusercontent.com/66107794/149273912-02bae8db-9452-4d4d-ad56-fa967f03f170.jpg)

But, lo and behold, nothing happens quite as expected in the IT world! After doing the restart to apply the change, I immediately noticed something strange; the 'Profile' page modules were no longer reporting properly;
![Freeze1](https://user-images.githubusercontent.com/66107794/149273992-ed31c76e-0138-48e9-b5d8-8c4b058f48f0.jpg)

Public page display was also frozen but songs were still playing through normally;
![Freeze2](https://user-images.githubusercontent.com/66107794/149274039-c7b00e0b-100b-4362-a50b-04fce5b53644.jpg)

Song queue was also weirding out;
![Freeze3](https://user-images.githubusercontent.com/66107794/149274071-60098539-4c4d-4820-9a89-f02cb13ec8d7.jpg)

But LS and the server were still running correctly as evidenced by ShoutCast page, player audio and our chatroom ;
![Shout](https://user-images.githubusercontent.com/66107794/149274094-4f0b0301-cf82-40a0-9467-618159fd4a28.jpg)

Neither rebooting the server nor restarting the broadcast helped with this. The only way out of this mess was to reset back to UTF-8, restart AzuraCast, then clear the song queue. Normal operation resumed fine after that.

I'm assuming this bug was never reported on before because we're probably one of the rare stations working both in French and ShoutCast...

Can someone give me the lowdown on what is happening here?

**Comment by toots** [comment]
The last time I looked at this, liquidsoap was receiving already corrupted strings with the [UTF8 replacement character](https://www.fileformat.info/info/unicode/char/fffd/index.htm), meaning that all information about this character was already lost and that something was off before the string reached liquidsoap.

I'd suggest to take a minute to analyze every step of your setup to better understand what strings are being used and in what encoding. Remember that encodings are not declared so each part of your system will assume a given encoding (latin1, ASCII, utf8) and process your strings as such. This means that if, for instance, a utf8 string hits a part of the system that assumes latin1 or ASCII, it will get corrupted.

* What are the encodings coming out of the file's metadata (if used)?
* What are the encoding coming out of the system's process call? What's the locale used in your server(s)?
* Etc...

Please also note that shoutcast and its ICY metadata protocol expects latin1, at least I checked. Typically, if you send UTF8 metadata to shoutcast without converting them to latin1 and then re-use the resulting string somewhere, your string will be corrupted.

**Reply by SpeedyCharly** [reply]
The metadata strings in question contain French characters é è ê ë à, etc. which can me formed differently on a computer depending on your keyboard configuration.  I, for example use an English keyboard set to US-International but most francos will have a French keyboard which may result in different coding being used.  mIRC users. for example, won't be able to download media with titles containing certain characters unless they turn off UTF-8 altogether.  I assume the problem is the same there...   

With over 200,000 title in my music collection it is an impossible task to revise all metadata. We've been operating this install for 2 years and I assure you that when our chat robot was reading the metadata directly from the LS/Azura stream we were invariably seeing the replacement characters. The robot is currently fetching the same metadata from the Shoutcast stream (/7.html) and posting correctly. Likewise our third-party radio player (RCAST) is posting the strings correctly.  All these are reading the same original metadata string so the question remains; what is different in LS.  

And why would me switching to ISO-8859-1 freeze a good portion of the Azuracast modules?

**Reply by toots** [reply]
I cannot comment on the reason switching the encoding on azuracast freezes the application though I suppose that it is could be because of some crash when processing/parsing the input.

Character encoding can be either mass-converted ahead of time or else you can try to auto-detect. I understand that you have a lot of data but, to paraphrase someone, "with great amount of data comes great amount of conversion". 🕷

String encoding is unrelated to how you produce letters on the keyboard. It is an _internal_ setting of computer systems. I can only, again, stress the need to document yourself a little bit more about the issue so that you are in a better position to understand and, potentially, fix it.

Again, I would suggest tracing the origin of the metadata that come into liquidsoap which, as I stated before, seem to be corrupted before. It could be because AzuraCast assumes UTF8 and you are feeding it non UTF8 data.

**Reply by toots** [reply]
Another thing: last I checked and, as stated above, it is likely the shoutcast server uses latin1 encoding for its strings, in which case, assuming this is true and your bot also expects latin1 characters, that would explain why it may be successful with shoutcast and not azuracast or liquidsoap which operate in UTF8.

**Reply by SpeedyCharly** [reply]
Called on the maintainer of our chatroom app and he assures me that chat and bot both run on UTF-8 protocol...

**Reply by SpeedyCharly** [reply]
FYI
https://github.com/ocaml/ocaml/issues/5348

**Reply by SpeedyCharly** [reply]
Don't know if you guys did anything on this issue lately, but I haven't seen UTF-8 or Chamomile errors for 2-3 days now!  Hosannas to the powers that be....
